Add "Navi Targeting Colors" options to Controller LED Color

sheepytina opened the pull request:
Makes the DualShock 4/DualSense controller Light Bar mirror whatever colour Navi is.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/996193932.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/996193933.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/996193934.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/996193936.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/996193937.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/996193938.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/996193939.zip)
<!--- section:artifacts:end -->

sheepytina committed:
Add a (not yet functional) "Navi" option.

sheepytina committed:
It works!

sheepytina committed:
Slightly better name for this menu option.

sheepytina committed:
Revert accidental commit of comments in z_en_elf.c

sheepytina commented on soh/src/overlays/actors/ovl_En_Elf/z_en_elf.c:
```diff
@@ -971,7 +971,7 @@ void func_80A03CF8(EnElf* this, PlayState* play) {
                         EnElf_SpawnSparkles(this, play, 16);
                     } else {
-                        if (distFromLinksHead > 100.0f) {
+                        if (distFromLinksHead > 100.0f) { // if Navi is too far away, she flies back.
```
I didn't intend to commit these comments and will push another commit that reverses this soon.

sheepytina commented on soh/include/z64.h:
```diff
@@ -2245,6 +2245,8 @@ typedef enum {
     LED_SOURCE_TUNIC_ORIGINAL,
     LED_SOURCE_TUNIC_COSMETICS,
     LED_SOURCE_HEALTH,
+    //LED_SOURCE_NAVI_ORIGINAL,
+    LED_SOURCE_NAVI,
```
I intend to go back and make it so Navi can either be the original colours or the Cosmetics Editor colours.

inspectredc commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1718,6 +1733,19 @@ Color_RGB8 GetColorForControllerLED() {
+                color.r = sNaviColorList[2].inner.r;
+                color.g = sNaviColorList[2].inner.g;
+                color.b = sNaviColorList[2].inner.b;
+            }
```
Would it be possible to have something along the lines of:

```cpp
switch (category) {
    case ACTORCAT_PLAYER:
        if (CVarGetInteger("gCosmetics.Navi_IdlePrimary.Changed", 0)) {
            color = CVarGetColor("gCosmetics.Navi_IdlePrimary.Value", defaultIdlePrimaryColor);
        } else {
            color.r = sNaviColorList[category].inner.r;
            color.g = sNaviColorList[category].inner.g;
            color.b = sNaviColorList[category].inner.b;
        }
        break;
    default:
        color.r = sNaviColorList[category].inner.r;
        color.g = sNaviColorList[category].inner.g;
        color.b = sNaviColorList[category].inner.b;
}
```

inspectredc replied:
Or possibly make custom cosmetic it's own option like with the tunic options

sheepytina replied:
Thank you for your suggestion! I was indeed planning on adding that but you helped a ton in pointing me in the right direction.

I wonder though, if it could be done even tidier if instead of a switch statement, it used an array of CVar names. With that said, the way it is now is probably a lot easier to read.

inspectredc replied:
Yeah I think there may be a way but it would always end up being much less readable than it is now

sheepytina committed:
Renaming and tweaks.

inspectredc commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -128,6 +128,21 @@ Color_RGB8 kokiriColor = { 0x1E, 0x69, 0x1B };
+    /* 0x04 */ Color_RGBA8 outer;
+} NaviColor; // size = 0x8
+
+NaviColor sNaviColorList[] = {
+    { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },         { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },
+    { { 255, 255, 255, 255 }, { 0, 0, 255, 0 } },     { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },
+    { { 150, 150, 255, 255 }, { 150, 150, 255, 0 } }, { { 255, 255, 0, 255 }, { 200, 155, 0, 0 } },
+    { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },         { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },
+    { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },         { { 255, 255, 0, 255 }, { 200, 155, 0, 0 } },
+    { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },         { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },
+    { { 0, 255, 0, 255 }, { 0, 255, 0, 0 } },
+};
```
Following on from the other review comment, you could then write out all the color options like defaultIdlePrimaryColor (or maybe defaultIdleInnerColor) and populate the array that way, e.g.

```cpp
NaviColor sNaviColorList[] = {
    { defaultInnerColor, defaultOuterColor},
     ... ,
    {defaultIdleInnerColor, defaultIdleOuterColor}, ...
}
```

inspectredc reviewed (commented):
Everything looks good! Left some comments for allowing for cosmetics (which i see now you're planning on adding now lol)

sheepytina committed:
Temporary fix(?) for compile error on non-Windows platforms.

sheepytina committed:
… Just pretend this particular commit doesn't exist. :OHYEAH:

sheepytina committed:
A more suitable fix. + Partial suggestions from review.

In-progress implementation of Cosmetics Navi colours.

sheepytina committed:
Convert relevant code to use Color_RGB8 sans alpha, matching the controller LED.

Defaults are now labeled for clarity.

sheepytina committed:
Revert back to Color_RGBA8. + Implement Cosmetics Navi colours.

Add Color conversion functions.

sheepytina committed:
Tidy comments.

sheepytina committed:
Changed mind yet a third time. Uses Color_RGB8 sans alpha again.

Uses CVarGetColor24 instead of CVarGetColor to drop alpha from cosmetics editor values, like how Tunic Colors does it.
This of course does require me to go with the prior idea of storing the Navi colours without alpha channels.

sheepytina committed:
Color type conversion functions removed

as they're no longer needed.

sheepytina committed:
Merge branch 'controller-led-is-navi' of https://github.com/sheepytina/Shipwright into controller-led-is-navi

sheepytina committed:
Tidy up commented out code once more.

sheepytina committed:
Fix a typo

inspectredc reviewed (approved):
Everything looks great!

inspectredc commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1718,6 +1746,46 @@ Color_RGB8 GetColorForControllerLED() {
+                        if (source == LED_SOURCE_NAVI_COSMETICS &&
+                            CVarGetInteger("gCosmetics.Navi_IdlePrimary.Changed", 0)) {
+                            color = CVarGetColor24("gCosmetics.Navi_IdlePrimary.Value", defaultIdleColor.inner);
+                            break;
```
Actually I think this should be outside the if statement

inspectredc replied:
The break that is

inspectredc replied:
Not sure how to best approach it, but since I believe the NPC cosmetic may not have changed but the enemy one may be changed, then this can lead to the NPC falling into the boss case and using that color. The simple fix is to just add the `color = LEDColorDefaultNaviColorList[category].inner;` at the end of every case followed by a `break;` or putting it in an else statement with the break outside. Not sure if theres something more elegant though

sheepytina replied:
> Actually I think this should be outside the if statement

~~Not necessarily. Break statements _can_ work like this; the idea is that it'll just continue to the default case if this IF statement doesn't resolve true. In practice, when using original colours instead of cosmetics ones.~~

~~If it's considered bad coding practice then I can rearrange it. Though if-breaks are used elsewhere in the code already.~~

EDIT: I replied to the first comment without refreshing the page to see the other ones that clarified what you meant.

sheepytina replied:
Although now I'm actually questioning myself on that.

But then if it didn't work, then cosmetics colours wouldn't work at all, and I have tested this.

inspectredc replied:
> Not necessarily. Break statements _can_ work like this; the idea is that it'll just continue to the default case if this IF statement doesn't resolve true. In practice, when using original colours instead of cosmetics ones.
> 
> If it's considered bad coding practice then I can rearrange it. Though if-breaks are used elsewhere in the code already.

Yeah that's what I thought of at first, however, I think it's possible for it to fail the ACTORCAT_NPC condition, but succeed the ACTORCAT_ENEMY condition. This is a result of them using different CVar 'has this cosmetic option changed'

sheepytina replied:
Alright, I see what you're saying now. It's the middle of the night here and I'm exhausted, but I'll take a look at it within 24 hours.

inspectredc commented:
I approve and then immediately notice something 🤦🏼

sheepytina committed:
Suggestion from code review. (Yeah nah you're totally right though.)

sheepytina commented:
A switch statement mishap has occurred. (5 dead, 12 injured)

I think I've fixed it now @inspectredc. Thank you for your patience.

sheepytina committed:
Correct indentation.

inspectredc reviewed (approved):
Yay

sheepytina committed:
A minor goof in the comments was bothering me.
